sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs

```diff
@@ -0,0 +1,170 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using System.Xml.Serialization;
+
+namespace Microsoft.PowerBI.Core.Cryptography
+{
+    /// <summary>
+    /// Helper class for encrypting and decrypting with asymmetric keys
+    /// </summary>
+    public static class AsymmetricKeyEncryptionHelper
+    {
+        private const int SegmentLength = 85;
+        private const int EncryptedLength = 128;
+
+        /// <summary>
+        /// Creates a new asymmetric key within the specified key container
+        /// </summary>
+        /// <param name="containerName"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static AsymmetricKey ProduceAsymmetricKey(string containerName)
+        {
+            if (string.IsNullOrEmpty(containerName)) throw new ArgumentNullException(nameof(containerName));
+
+            var cspParams = new CspParameters { KeyContainerName = containerName };
+
+            // delete any old keys associated with the existing container
+            using (var rsa = new RSACryptoServiceProvider(cspParams))
+            {
+                rsa.PersistKeyInCsp = false;
+                rsa.Clear();
+            }
+
+            using (var rsa = new RSACryptoServiceProvider(cspParams))
+            {
+                // The public key will be saved in PBI service when gateway register
+                // The private key will ONLY be stored on the GW machine
+                // We don't directly handle private key, but we do retrieve the public key as part of the step
+                var publicKey = rsa.ExportParameters(false);
+
+                return new AsymmetricKey
+                {
+                    KeyContainerName = containerName,
+                    PublicKey = ConvertKeyToString(publicKey),
+                };
+            }
+        }
+
+        /// <summary>
+        /// Encrypts the text from the keys in the specified container
+        /// </summary>
+        /// <param name="plainText">The text to encrypt</param>
+        /// <param name="containerName">The key container name</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static string Encrypt(string plainText, string containerName)
```

Comment:
This is copied directly from the DM encryption method.  It appears it is non-standard to begin with and we'll need to conform to what DM expects